api/src/server.rs

```diff
@@ -817,6 +832,142 @@ where
         let stop = stop_builder.with_max_event().build_fencepost();
         Ok((start, stop))
     }
+
+    async fn get_stream_state(
+        &self,
+        stream_id: String,
+    ) -> Result<ceramic_api_server::StreamsStreamIdGetResponse, ErrorResponse> {
+        let stream_id = StreamId::from_str(&stream_id)
+            .map_err(|err| ErrorResponse::new(format!("invalid stream_id: {err}")))?;
+        let state_batch = self
+            .pipeline
+            .table("doc_state")
+            .await
+            .map_err(|err| ErrorResponse::new(format!("doc_state table not found: {err}")))?
+            .select(vec![
+                col("stream_cid"),
+                col("event_cid"),
+                col("dimensions"),
+                col("controller"),
+                col("state"),
+                col("index"),
+            ])
+            .map_err(|err| ErrorResponse::new(format!("failed to select: {err}")))?
+            .filter(col("stream_cid").eq(lit(stream_id.cid.to_bytes())))
+            .map_err(|err| ErrorResponse::new(format!("failed to filter: {err}")))?
+            .aggregate(
+                vec![col("stream_cid"), col("controller")],
+                vec![
+                    last_value(vec![col("state")])
+                        .order_by(vec![col("index").sort(true, true)])
+                        .build()
+                        .map_err(|err| {
+                            ErrorResponse::new(format!(
+                                "failed to define last_value state query: {err}"
+                            ))
+                        })?
+                        .alias("state"),
+                    last_value(vec![col("event_cid")])
+                        .order_by(vec![col("index").sort(true, true)])
```

Comment:
kinda sucks that you have to do this sort statement multiple times for each column.  I'm surprised you can't just do a sort of al the rows once, then select the columns from the selected row?

Reply:
same, I was annoyed as well. Should not affect performance but its quite verbose.